### PR TITLE
test: kernel: smp: accommodate more cores

### PR DESCRIPTION
**Pull request by evgeniy-paltsev**
Now Zephyr allow to have up to 12 CPU cores (see [MP_MAX_NUM_CPUS value](https://github.com/zephyrproject-rtos/zephyr/blob/859cc2f1feec93132e06753278be02889418333f/kernel/Kconfig.smp#L51)), however `tests/kernel/smp/` logic hasn't been adjusted for that - so it would always fail on such platforms.

Let's fix `tests/kernel/smp/` so it can run on more CPU cores.

While I'm at it - also make `test_coop_resched_threads` test case logic more robust - so it won't rely on the combination of three timeouts (and actual time required to spawn all threads)

### CODE REVIEW

**Review by abrodkin — DISMISSED** [review]

**Review by abrodkin — APPROVED** [review]

File: tests/kernel/smp/src/main.c [hunk 1]
```diff
@@ -224,6 +230,17 @@ static void thread_entry_fn(void *p1, void *p2, void *p3)
 	}
 }
 
+static void thread_entry_inf_fn(void *p1, void *p2, void *p3)
```
Comment (andyross) [inline tests/kernel/smp/src/main.c:233]: So... I do not disagree that this is an improvement to the design.  But... it's test code.  And the earlier code wasn't actually wrong, it was just clumsy.  But here's the important part: it was clumsy becuase it was making assumptions that the five 50ms k_busy_wait() calls in the function would not interfere with the smaller delay in test_coop_resched_thread().  And sure: that k_busy_wait() isn't supposed to interfere or be slow or wrong.  And it isn't.  So it's needless.  **In app code**.  But in the test, what's happened is that you've actually removed coverage of k_busy_wait().

Basically, my feeling is that "bad test code is only bad if it's invalid".  Merely "testing the wrong thing" does more good than harm.  There have been a lot of edge cases found via weird test code, historically.

tl;dr: Can you remove this patch?

File: tests/kernel/smp/src/main.c [hunk 2]
```diff
@@ -208,7 +214,12 @@ static void thread_entry_fn(void *p1, void *p2, void *p3)
 	tinfo[thread_num].executed  = 1;
 	tinfo[thread_num].cpu_id = curr_cpu();
 
-	while (count++ < 5) {
+	/* We use this thread function with spawn_threads helper which increase delay before each
+	 * new thread is spawned. So we need take into account the amount of threads we are going to
+	 * spawn to wait enough time before exiting the thread (alternatively we could just increase
+	 * wait time for all configurations)
+	 */
+	while (count++ < (5 + arch_num_cpus())) {
```
Comment (andyross) [inline tests/kernel/smp/src/main.c:228]: MISRA police are going to have you jailed for that side effect in the expression

File: tests/kernel/smp/prj.conf [hunk 3]
```diff
@@ -2,3 +2,5 @@ CONFIG_ZTEST=y
 CONFIG_SMP=y
 CONFIG_TRACE_SCHED_IPI=y
 CONFIG_POLL=y
+CONFIG_NUM_PREEMPT_PRIORITIES=32
+CONFIG_NUM_COOP_PRIORITIES=32
```
Comment (andyross) [inline tests/kernel/smp/prj.conf:6]: Seems like this is more than you need?  Also pedantry: the idle thread is not included in the preemptive count (for dumb historical reasons -- it used to be a special cased priority), so this is going to create a configuration with 65 total priorities, not 64, and thus require an extra word of storage in the MULTIQ scheduler bitmask.

**Review by andyross — CHANGES_REQUESTED** [review]
SMP test looks great.  But IMHO the patch to the test is a little too improved.

File: tests/kernel/smp/src/main.c (hunk 2 above)
Comment (evgeniy-paltsev) [reply tests/kernel/smp/src/main.c:228]: As I remember we don't enforce them in the test code, don't we?

File: tests/kernel/smp/src/main.c (hunk 1 above)
Comment (evgeniy-paltsev) [reply tests/kernel/smp/src/main.c:233]: First of all, about the rationale:

in the original test case we are dealing with three timeouts:

(1) Time we spend in thread_entry_fn:
https://github.com/zephyrproject-rtos/zephyr/blob/64894e8798d41bf2b6c8c3e36def394ba0a7bd82/tests/kernel/smp/src/main.c#L211-L212

(2) Time we wait till the threads spawn on other cores:
https://github.com/zephyrproject-rtos/zephyr/blob/64894e8798d41bf2b6c8c3e36def394ba0a7bd82/tests/kernel/smp/src/main.c#L336

(3) Time we delay creation of new thread (this one is even different for each thread)
https://github.com/zephyrproject-rtos/zephyr/blob/64894e8798d41bf2b6c8c3e36def394ba0a7bd82/tests/kernel/smp/src/main.c#L245-L253

On top of that there is some time we need to spawn new thread itself (4).

So we have multiple constraints which need to be true to have the test case logic correct:
max(3) < (1)
max(3) + (4) < (2)
min(3) + (4) + (1) > (2)
(2) < (1)
Note that 3 would differ for each thread and it maximum value depend on number of threads (number of cores)

The test case idea is to check that new coop thread with higher priority created won't preempt existing coop thread, and such dance with delays doesn't look robust at all.


> what's happened is that you've actually removed coverage of k_busy_wait()

That shouldn't decrease coverage of k_busy_wait:
 - we have tests for k_busy_wait (and even in `tests/kernel/smp` we still use k_busy_wait more than 13 times)
 - we don't call it with some specific parameter combinations
 - we don't call it in some specific context

So, I can remove it (or move it to separate PR), however I'm not really convinced here.

Comment (andyross) [reply tests/kernel/smp/src/main.c:233]: This is making me less confident in your solution and not more.  Now I need to dig around and figure out exactly why that very large (250ms!!!!) delay is somehow not enough to spawn 12 threads...

Basically: are you 100% absolutely sure you aren't just working around a platform bug with k_busy_wait(), maybe because it's saturating a bus or timer device or something and preventing other CPUs from working?

File: tests/kernel/smp/src/main.c [hunk 4]
```diff
@@ -25,6 +25,12 @@
 #define THREAD_DELAY 1
 #define SLEEP_MS_LONG ((int)(15000 * RUN_FACTOR))
 
+/* We iterate priorities in spawn_threads() - so we need to have enough priorities */
+#define SPAWN_THREADS_MAX_PRIO 16
+BUILD_ASSERT(SPAWN_THREADS_MAX_PRIO >= CONFIG_MP_MAX_NUM_CPUS);
```
Comment (cfriedt) [inline tests/kernel/smp/src/main.c:30]: This is a bit of a shortcut. Alternatively, make a Kconfig range option local to this test, where the lower bound is here, and the upper bound is one of the values below.
```suggestion
#define SPAWN_THREADS_MAX_PRIO CONFIG_MP_MAX_NUM_CPUS
```
